sdk/metric/sdk.go

```diff
@@ -186,30 +149,49 @@ func (i *instrument) acquireHandle(ls *labels) *record {
 	if actual, ok := i.meter.current.Load(mk); ok {
 		// Existing record case, only one allocation so far.
 		rec := actual.(*record)
-		atomic.AddInt64(&rec.refcount, 1)
-		return rec
+		if rec.refMapped.ref() {
+			// At this moment it is guaranteed that the entry is in
+			// the map and will not be removed.
+			return rec
+		}
+		// This entry is nolonger mapped, try to add a new entry.
 	}
 
 	// There's a memory allocation here.
 	rec := &record{
-		labels:         ls,
-		descriptor:     i.descriptor,
-		refcount:       1,
-		collectedEpoch: -1,
-		modifiedEpoch:  0,
-		recorder:       i.meter.batcher.AggregatorFor(i.descriptor),
+		labels:     ls,
+		descriptor: i.descriptor,
+		refMapped:  refcountMapped{value: 2},
+		recorder:   i.meter.batcher.AggregatorFor(i.descriptor),
 	}
 
-	// Load/Store: there's a memory allocation to place `mk` into
-	// an interface here.
-	if actual, loaded := i.meter.current.LoadOrStore(mk, rec); loaded {
-		// Existing record case.
-		rec = actual.(*record)
-		atomic.AddInt64(&rec.refcount, 1)
-		return rec
+	for {
+		// Load/Store: there's a memory allocation to place `mk` into
+		// an interface here.
+		if actual, loaded := i.meter.current.LoadOrStore(mk, rec); loaded {
+			// Existing record case. Cannot change rec here because if fail
+			// will try to add rec again to avoid new allocations.
+			oldRec := actual.(*record)
+			if oldRec.refMapped.ref() {
+				// At this moment it is guaranteed that the entry is in
+				// the map and will not be removed.
+				return oldRec
+			}
+			// This loaded entry is marked as unmapped (so Collect will remove
+			// it from the map immediately), try again - this is a busy waiting
+			// strategy to wait until Collect() removes this entry from the map.
```

Comment:
It does not wait an our, the most waiting time is when the kernel deschedules the collector thread (goroutine) between the moment tryUnmap succeed and Delete operation in the Collect() function, so it can wait until the thread is rescheduled (it is not an our for sure).

Reply:
Ah, yes, now I can see that this can only happen when Collect unmapped the entry, but haven't yet removed it from the `current` map. I suppose that busy looping here is just fine.

Reply:
That is my thought as well, it shouldn't be a problem in practice.

Reply:
Should we put a call to `runtime.Gosched()` in, then? There's similar logic in the prometheus metrics client, where it does this.

Reply:
Also I'd like to at least put in a firm comment saying that a metric operation could block as a result of this. We think this will not be a problem.  The existing code, which may have had another bug, at least didn't have a blocking problem. Still I like to see simpler code.

Reply:
@jmacd your statement is incorrect, the code was blocking before as well (with a much higher chance of actually blocking because of a global map between instruments) because of the sync.Map.LoadOrStore (look at line 209 https://golang.org/src/sync/map.go?s=6882:6965#L189).

If you want to add that message feel free to add it, but would suggest that the primary cause of blocking is actually sync.Map :)

Reply:
Also any place where you have a cas inside a for loop (and there were at least 2 on the critical path one for addPrimary and one for addReclaimed) can be seen as a blocking scenario because of potential contention and one goroutine can simply spin for a very long time there.

Reply:
I accept! :-)